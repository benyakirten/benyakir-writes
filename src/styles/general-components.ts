import styled from "styled-components";

import { fadeIn, slideInLeft, slideInRight } from "./animations";
import { media } from "./queries";
import {
	FONT_SIZE_MD,
	SIZE_MD,
	TRANSITION_EXTRA_SLOW,
	TRANSITION_NORMAL,
	SANS_SERIF_FONT,
	SIZE_SM,
	SIZE_XS,
	SIZE_LG,
	FONT_SIZE_XXS,
	FONT_XXXL,
	FONT_XXL,
	FONT_XL,
	FONT_MD,
	FONT_LG,
} from "./variables";

import { hoverUnderline, textClipWithHoverTransition } from "./style-mixins";

export const FadeIn = styled.div<{ duration?: string; delay?: string }>`
  opacity: 0;
  animation: ${fadeIn}
    ${(props) => (props.duration ? props.duration : TRANSITION_NORMAL)}
    ${(props) => (props.delay ? props.delay : "0ms")} ease forwards;
`;

export const SlideInLeft = styled.div<{ duration?: string; delay?: string }>`
  opacity: 0;
  transform: translateX(-40vw);
  animation: ${slideInLeft}
    ${(props) => (props.duration ? props.duration : TRANSITION_EXTRA_SLOW)}
    ${(props) => (props.delay ? props.delay : "0ms")} ease forwards;
`;

export const SlideInRight = styled.div<{ duration?: string; delay?: string }>`
  opacity: 0;
  transform: translateX(40vw);
  animation: ${slideInRight}
    ${(props) => (props.duration ? props.duration : TRANSITION_EXTRA_SLOW)}
    ${(props) => (props.delay ? props.delay : "0ms")} ease forwards;
`;

export const Grouping = styled.section<{ marginVertical?: string }>`
  position: relative;
  margin-top: ${(props) => (props.marginVertical ? props.marginVertical : "0")};
  margin-bottom: ${(props) =>
		props.marginVertical ? props.marginVertical : "4rem"};
`;

// TODO: Make the colors look good
export const LeadHeading = styled.h1`
  position: relative;

  margin-bottom: ${SIZE_MD};

  ${FONT_XXXL};

  ${(props) => textClipWithHoverTransition(props.theme.header.startColor, props.theme.header.endColor)}

  ${media.phone} {
    ${FONT_XXL};
  }
`;

export const Subtitle = styled.h2`
  position: relative;

  margin-bottom: ${SIZE_MD};

  ${FONT_XL};

  ${(props) => textClipWithHoverTransition(props.theme.header.startColor, props.theme.header.endColor)}

  ${media.phone} {
    ${FONT_LG};
  }
`;

export const BigParagraph = styled.p`
  ${FONT_XL};
`;

export const WpContentDescription = styled.div<{ fontSize?: string }>`
  ${FONT_MD};

  overflow: hidden;
  text-overflow: ellipsis;

  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;

  a:link,
  a:visited {
    position: relative;

    color: ${(props) => props.theme.base.textColor};

    overflow: hidden;
    white-space: nowrap;

    &::after {
      content: '';

      position: absolute;
      left: 0;
      bottom: 0;

      height: 2px;
      width: 100%;

      background-color: ${(props) => props.theme.base.background};

      transition: transform ${TRANSITION_NORMAL} ease;
      transform-origin: left;
      transform: scaleX(0);
    }

    &:hover {
      &::after {
        transform: scaleX(1);

        ${media.reducedMotion} {
          transform: scaleX(0);
        }
      }
    }
  }
`;

export const Card = styled.div`
  border-radius: ${SIZE_XS};
  border: 1px solid ${(props) => props.theme.base.border};

  padding: ${SIZE_SM};
  background-color: ${(props) => props.theme.base.background};
`;

export const Box = styled.div`
  display: flex;
  align-items: center;
  gap: ${SIZE_MD};
`;

export const FlatBox = styled(Box)`
  margin: 0;
`;

// https://stackoverflow.com/questions/24933430/img-src-svg-changing-the-styles-with-css
export const SVGContainer = styled.div<{
	src: string;
	alt: string;
	size?: string;
}>`
  position: relative;

  background-color: ${(props) => props.theme.base.textColor};
  mask: url(${(props) => props.src}) no-repeat center;

  height: ${(props) => props.size ?? SIZE_MD};
  width: ${(props) => props.size ?? SIZE_MD};
`;

export const WpContent = styled.div`
  ${FONT_LG};

  ${media.phone} {
    ${FONT_MD};
  }

  a:link,
  a:visited {
    ${(props) => hoverUnderline(props.theme.base.link)}
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    font-family: ${SANS_SERIF_FONT};
    padding: ${SIZE_MD} 0;
    ${(props) => textClipWithHoverTransition(props.theme.header.startColor, props.theme.header.endColor)}
  }


  ul,
  ol,
  li {
    margin-left: ${SIZE_SM};
  }

  ul:has(li > a), ol:has(li > a) {
    list-style-type: disc;
  }

  p {
    margin: ${SIZE_MD} 0;
  }

  code {
    font-family: monospace;
    font-size: ${FONT_SIZE_MD};
  }
`;

export const Page = styled.div`
  overflow-y: auto;
  height: 100vh;
`;

export const PageContents = styled.div`
  padding: 2rem 4rem;
  ${media.tablet} {
    padding: 1rem 4rem;
  }
`;

export const PillContainer = styled.div`
	display: flex;
	align-items: center;
	justify-content: space-between;

  white-space: nowrap;
	
  height: min-content;
  width: min-content;

	font-family: ${SANS_SERIF_FONT};
  font-size: ${FONT_SIZE_XXS};
  text-transform: capitalize;
  
  border-radius: ${SIZE_LG};
  color: ${(props) => props.theme.base.textColor};
  background-color: ${(props) => props.theme.pill.background};

	padding: ${SIZE_XS} ${SIZE_SM};

  ${media.phone} {
    padding: ${SIZE_XS};
  }
`;

export const PaginatedPageContents = styled.div`
  padding: 2rem 18rem 4rem 1rem;
  ${media.phone} {
    padding: 2rem 0.5rem 4rem 0.5rem;
  }
`;

export const NormalPageContents = styled.div`
  padding: 2rem 1rem;
  ${media.phone} {
    padding: 2rem 0.5rem;
  }
`;

export const TemplateHeaderTitle = styled(LeadHeading)`
  margin-bottom: 0;
`;

export const TemplateContent = styled.div`
  margin: ${SIZE_MD} 0;
`;

export const TemplateHeaderContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${SIZE_XS};
  margin-bottom: ${SIZE_SM};
`;
