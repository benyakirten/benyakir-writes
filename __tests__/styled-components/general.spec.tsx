import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";

import * as General from "@Styles/general-components";

describe("general styled components", () => {
	describe("Transition components", () => {
		describe("FadeIn", () => {
			it("should accept a duration and delay prop that determines the animations duration and delay of the component", () => {
				const DELAY_1 = "300ms";
				const DURATION_1 = "1000ms";
				const testComp1 = renderer
					.create(<General.FadeIn delay={DELAY_1} duration={DURATION_1} />)
					.toJSON();

				expect(testComp1).toMatchSnapshot();
				expect(testComp1).toHaveStyleRule(
					"animation",
					`dMKEGd ${DURATION_1} ${DELAY_1} ease forwards`,
				);

				const DELAY_2 = "600ms";
				const DURATION_2 = "2000ms";
				const testComp2 = renderer
					.create(<General.FadeIn delay={DELAY_2} duration={DURATION_2} />)
					.toJSON();

				expect(testComp2).toMatchSnapshot();
				expect(testComp2).toHaveStyleRule(
					"animation",
					`dMKEGd ${DURATION_2} ${DELAY_2} ease forwards`,
				);
			});

			it("should have a default value for the delay and duration if none is provided", () => {
				const testComp = renderer.create(<General.FadeIn />).toJSON();
				expect(testComp).toMatchSnapshot();
				expect(testComp).toHaveStyleRule(
					"animation",
					`dMKEGd 1600ms 0ms ease forwards`,
				);
			});
		});

		describe("SlideInLeft", () => {
			it("should accept a duration and delay prop that determines the animations duration and delay of the component", () => {
				const DELAY_1 = "300ms";
				const DURATION_1 = "1000ms";
				const testComp1 = renderer
					.create(<General.SlideInLeft delay={DELAY_1} duration={DURATION_1} />)
					.toJSON();

				expect(testComp1).toMatchSnapshot();
				expect(testComp1).toHaveStyleRule(
					"animation",
					`hOeomY ${DURATION_1} ${DELAY_1} ease forwards`,
				);

				const DELAY_2 = "600ms";
				const DURATION_2 = "2000ms";
				const testComp2 = renderer
					.create(<General.SlideInLeft delay={DELAY_2} duration={DURATION_2} />)
					.toJSON();

				expect(testComp2).toMatchSnapshot();
				expect(testComp2).toHaveStyleRule(
					"animation",
					`hOeomY ${DURATION_2} ${DELAY_2} ease forwards`,
				);
			});

			it("should have a default value for the delay and duration if none is provided", () => {
				const testComp = renderer.create(<General.SlideInLeft />).toJSON();
				expect(testComp).toMatchSnapshot();
				expect(testComp).toHaveStyleRule(
					"animation",
					`hOeomY 4s 0ms ease forwards`,
				);
			});

			describe("SlideInRight", () => {
				it("should accept a duration and delay prop that determines the animations duration and delay of the component", () => {
					const DELAY_1 = "300ms";
					const DURATION_1 = "1000ms";
					const testComp1 = renderer
						.create(
							<General.SlideInRight delay={DELAY_1} duration={DURATION_1} />,
						)
						.toJSON();

					expect(testComp1).toMatchSnapshot();
					expect(testComp1).toHaveStyleRule(
						"animation",
						`dcBvEg ${DURATION_1} ${DELAY_1} ease forwards`,
					);

					const DELAY_2 = "600ms";
					const DURATION_2 = "2000ms";
					const testComp2 = renderer
						.create(
							<General.SlideInRight delay={DELAY_2} duration={DURATION_2} />,
						)
						.toJSON();

					expect(testComp2).toMatchSnapshot();
					expect(testComp2).toHaveStyleRule(
						"animation",
						`dcBvEg ${DURATION_2} ${DELAY_2} ease forwards`,
					);
				});

				it("should have a default value for the delay and duration if none is provided", () => {
					const testComp = renderer.create(<General.SlideInRight />).toJSON();
					expect(testComp).toMatchSnapshot();
					expect(testComp).toHaveStyleRule(
						"animation",
						`dcBvEg 4s 0ms ease forwards`,
					);
				});
			});
		});
	});

	describe("Text components", () => {
		// This errors out - I need to use the css`` function somehow, but I'm not sure
		// And I'm not sure why it doesn't work when the following one does work

		// describe('Subtitle', () => {
		//     it('should set the text-decoration property to none based on the noUnderline prop', () => {
		//         const comp1 = renderer.create(<General.Subtitle noUnderline />).toJSON()
		//         expect(comp1).toMatchSnapshot()
		//         expect(comp1).toHaveStyleRule('text-decoration', 'none')

		//         const comp2 = renderer.create(<General.Subtitle noUnderline={false} />).toJSON()
		//         expect(comp2).toMatchSnapshot()
		//         expect(comp2).toHaveStyleRule('text-decoration', 'underline')
		//     })

		//     it('should set the text-decoration by default to underline if the no Underline prop is not set', () => {
		//         const comp = renderer.create(<General.Subtitle />).toJSON()
		//         expect(comp).toMatchSnapshot()
		//         expect(comp).toHaveStyleRule('text-decoration', 'underline')
		//     })
		// })

		describe("SubHeading", () => {
			it("should set the text-decoration property to none based on the noUnderline prop", () => {
				const comp1 = renderer
					.create(<General.SubHeading noUnderline />)
					.toJSON();
				expect(comp1).toMatchSnapshot();
				expect(comp1).toHaveStyleRule("text-decoration", "none");

				const comp2 = renderer
					.create(<General.SubHeading noUnderline={false} />)
					.toJSON();
				expect(comp2).toMatchSnapshot();
				expect(comp2).toHaveStyleRule("text-decoration", "underline");
			});

			it("should set the text-decoration by default to underline if the no Underline prop is not set", () => {
				const comp = renderer.create(<General.SubHeading />).toJSON();
				expect(comp).toMatchSnapshot();
				expect(comp).toHaveStyleRule("text-decoration", "underline");
			});
		});

		describe("BigParagraph", () => {
			it("should set the margin-right property to the marginRight prop passed in", () => {
				const comp = renderer
					.create(<General.BigParagraph marginRight="12rem" />)
					.toJSON();
				expect(comp).toMatchSnapshot();
				expect(comp).toHaveStyleRule("margin-right", "12rem");
			});

			it("should set a default margin-right of 0 if there is no marginRight prop passed in", () => {
				const comp = renderer.create(<General.BigParagraph />).toJSON();
				expect(comp).toMatchSnapshot();
				expect(comp).toHaveStyleRule("margin-right", "0");
			});
		});
	});

	describe("Other components", () => {
		describe("HoverableContainer", () => {
			it("should set the height and width peropties based on the props passed in", () => {
				const comp = renderer
					.create(<General.HoverableContainer height={400} width={800} />)
					.toJSON();
				expect(comp).toMatchSnapshot();
				expect(comp).toHaveStyleRule("height", "400px");
				expect(comp).toHaveStyleRule("width", "800px");
			});

			it("should set a default height and width if there is no marginRight prop passed in", () => {
				const comp = renderer.create(<General.HoverableContainer />).toJSON();
				expect(comp).toMatchSnapshot();
				expect(comp).toHaveStyleRule("height", "200px");
				expect(comp).toHaveStyleRule("width", "134px");
			});
		});
	});
});
