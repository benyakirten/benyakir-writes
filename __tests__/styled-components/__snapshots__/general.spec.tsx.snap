// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`general styled components Other components HoverableContainer should set a default height and width if there is no marginRight prop passed in 1`] = `
.c0 {
  height: 200px;
  width: 134px;
  border: 2px solid #05491F;
  overflow: hidden;
}

<div
  className="c0"
/>
`;

exports[`general styled components Other components HoverableContainer should set the height and width peropties based on the props passed in 1`] = `
.c0 {
  height: 400px;
  width: 800px;
  border: 2px solid #05491F;
  overflow: hidden;
}

<div
  className="c0"
  height={400}
  width={800}
/>
`;

exports[`general styled components Text components BigParagraph should set a default margin-right of 0 if there is no marginRight prop passed in 1`] = `
.c0 {
  font-size: 2.4rem;
  margin-top: 0;
  margin-bottom: 0;
  margin-right: 0;
}

<p
  className="c0"
/>
`;

exports[`general styled components Text components BigParagraph should set the margin-right property to the marginRight prop passed in 1`] = `
.c0 {
  font-size: 2.4rem;
  margin-top: 0;
  margin-bottom: 0;
  margin-right: 12rem;
}

<p
  className="c0"
/>
`;

exports[`general styled components Text components SubHeading should set the text-decoration by default to underline if the no Underline prop is not set 1`] = `
.c0 {
  font-family: 'Faustina',serif;
  font-size: 2.4rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin-bottom: 0.5rem;
}

@media (max-width:768px) {
  .c0 {
    font-size: 1.8rem;
  }
}

<h3
  className="c0"
/>
`;

exports[`general styled components Text components SubHeading should set the text-decoration property to none based on the noUnderline prop 1`] = `
.c0 {
  font-family: 'Faustina',serif;
  font-size: 2.4rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-bottom: 0.5rem;
}

@media (max-width:768px) {
  .c0 {
    font-size: 1.8rem;
  }
}

<h3
  className="c0"
/>
`;

exports[`general styled components Text components SubHeading should set the text-decoration property to none based on the noUnderline prop 2`] = `
.c0 {
  font-family: 'Faustina',serif;
  font-size: 2.4rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin-bottom: 0.5rem;
}

@media (max-width:768px) {

}

@media (max-width:768px) {
  .c0 {
    font-size: 1.8rem;
  }
}

<h3
  className="c0"
/>
`;

exports[`general styled components Transition components FadeIn should accept a duration and delay prop that determines the animations duration and delay of the component 1`] = `
.c0 {
  opacity: 0;
  -webkit-animation: dMKEGd 1000ms 300ms ease forwards;
  animation: dMKEGd 1000ms 300ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components FadeIn should accept a duration and delay prop that determines the animations duration and delay of the component 2`] = `
.c0 {
  opacity: 0;
  -webkit-animation: dMKEGd 2000ms 600ms ease forwards;
  animation: dMKEGd 2000ms 600ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components FadeIn should have a default value for the delay and duration if none is provided 1`] = `
.c0 {
  opacity: 0;
  -webkit-animation: dMKEGd 1600ms 0ms ease forwards;
  animation: dMKEGd 1600ms 0ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components SlideInLeft SlideInRight should accept a duration and delay prop that determines the animations duration and delay of the component 1`] = `
.c0 {
  opacity: 0;
  -webkit-transform: translateX(40vw);
  -ms-transform: translateX(40vw);
  transform: translateX(40vw);
  -webkit-animation: dcBvEg 1000ms 300ms ease forwards;
  animation: dcBvEg 1000ms 300ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components SlideInLeft SlideInRight should accept a duration and delay prop that determines the animations duration and delay of the component 2`] = `
.c0 {
  opacity: 0;
  -webkit-transform: translateX(40vw);
  -ms-transform: translateX(40vw);
  transform: translateX(40vw);
  -webkit-animation: dcBvEg 2000ms 600ms ease forwards;
  animation: dcBvEg 2000ms 600ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components SlideInLeft SlideInRight should have a default value for the delay and duration if none is provided 1`] = `
.c0 {
  opacity: 0;
  -webkit-transform: translateX(40vw);
  -ms-transform: translateX(40vw);
  transform: translateX(40vw);
  -webkit-animation: dcBvEg 4s 0ms ease forwards;
  animation: dcBvEg 4s 0ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components SlideInLeft should accept a duration and delay prop that determines the animations duration and delay of the component 1`] = `
.c0 {
  opacity: 0;
  -webkit-transform: translateX(-40vw);
  -ms-transform: translateX(-40vw);
  transform: translateX(-40vw);
  -webkit-animation: hOeomY 1000ms 300ms ease forwards;
  animation: hOeomY 1000ms 300ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components SlideInLeft should accept a duration and delay prop that determines the animations duration and delay of the component 2`] = `
.c0 {
  opacity: 0;
  -webkit-transform: translateX(-40vw);
  -ms-transform: translateX(-40vw);
  transform: translateX(-40vw);
  -webkit-animation: hOeomY 2000ms 600ms ease forwards;
  animation: hOeomY 2000ms 600ms ease forwards;
}

<div
  className="c0"
/>
`;

exports[`general styled components Transition components SlideInLeft should have a default value for the delay and duration if none is provided 1`] = `
.c0 {
  opacity: 0;
  -webkit-transform: translateX(-40vw);
  -ms-transform: translateX(-40vw);
  transform: translateX(-40vw);
  -webkit-animation: hOeomY 4s 0ms ease forwards;
  animation: hOeomY 4s 0ms ease forwards;
}

<div
  className="c0"
/>
`;
